near-sdk/src/private.rs

```diff
@@ -0,0 +1,61 @@
+use crate::{Balance, Gas, Promise, PromiseOrValue};
+
+pub struct FunctionCallBuilder {
+    promise: Option<Promise>,
+    method_name: String,
+    args: Vec<u8>,
+    amount: Balance,
+    gas: Option<Gas>,
+}
+
+impl FunctionCallBuilder {
```

Comment:
I suggest we don't introduce new structure `FunctionCallBuilder` and just implement these methods to `Promise` class itself. This will avoid having `into_promise` method that all users need to call, which would make API simpler.

AFAIK there are two ways the builder pattern can be implemented:
* With one struct/class. `FinalStruct` that has methods `add_field1(FinalStruct,...) -> FinalStruct`;
* With two structs/classes. `Builder` that has methods `add_field1(Builder,...) -> Builder` and `finalize(Builder)->FinalStruct`.

AFAIU the second one is typically used when `FinalStruct` is heavy (e.g. building an RPC server), so I think we should go with the first one.

Reply:
This would require having `with_gas` and `with_balance` as part of the `Promise` API, when you would not be able to know which part of the promise you were specifying gas/balance for. There needs to be some compile-time restriction to scope this function call builder from the Promise, else this becomes a really janky API IMO.

Because you can't `Deref` into an owned type and because the promise builder pattern is `<function>(self,.. ) -> Self` instead of references to `self`, this is why the into_promise is required.